internal/cache/file/downloader/job.go

```diff
@@ -197,6 +203,101 @@ func (job *Job) updateFileInfoCache() (err error) {
 	return
 }
 
+// downloadObjectAsync downloads the backing GCS object into a file as part of
+// file cache using NewReader method of gcs.Bucket.
+//
+// Note: There can only be one async download running for a job at a time.
+// Acquires and releases LOCK(job.mu)
+func (job *Job) downloadObjectAsync() {
+	// Create and open cache file for writing object into it.
+	cacheFile, err := util.CreateFile(job.fileSpec, os.O_WRONLY)
+	defer func(cacheFile *os.File) {
+		err = cacheFile.Close()
+		if err != nil {
+			err = fmt.Errorf("downloadObjectAsync: error while closing cache file: %v", err)
+			job.failWhileDownloading(err)
+		}
+	}(cacheFile)
+	if err != nil {
+		err = fmt.Errorf("downloadObjectAsync: error in creating cache file: %v", err)
+		job.failWhileDownloading(err)
+	}
+
+	var newReader io.ReadCloser
+	var start, end, sequentialReadSize, newReaderLimit int64
+	end = int64(job.object.Size)
+	sequentialReadSize = int64(job.sequentialReadSizeMb) * MiB
+
+	for {
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
+			if start < end {
+				if newReader == nil {
+					newReaderLimit = min(start+sequentialReadSize, end)
+					newReader, err = job.bucket.NewReader(
+						job.cancelCtx,
+						&gcs.ReadObjectRequest{
+							Name:       job.object.Name,
+							Generation: job.object.Generation,
+							Range: &gcs.ByteRange{
+								Start: uint64(start),
+								Limit: uint64(newReaderLimit),
+							},
+							ReadCompressed: job.object.HasContentEncodingGzip(),
+						})
+					if err != nil {
+						err = fmt.Errorf(fmt.Sprintf("downloadObjectAsync: error in creating NewReader with start %d and limit %d: %v", start, newReaderLimit, err))
+						job.failWhileDownloading(err)
+						return
+					}
+				}
+
+				maxRead := min(end-start, ReadChunkSize)
+				_, err = cacheFile.Seek(start, 0)
+				if err != nil {
+					err = fmt.Errorf(fmt.Sprintf("downloadObjectAsync: error while seeking file handle, seek %d: %v", start, err))
+					job.failWhileDownloading(err)
+					return
+				}
+
+				// copy the contents from NewReader to cache file.
+				_, readErr := io.CopyN(cacheFile, newReader, maxRead)
+				if readErr != nil && readErr != io.EOF {
+					err = fmt.Errorf("downloadObjectAsync: error at the time of copying content to cache file %v", readErr)
+					job.failWhileDownloading(err)
+					return
+				}
+				start += maxRead
+				if readErr == io.EOF {
+					newReader = nil
+				}
+
+				job.mu.Lock()
+				job.status.Offset = start
```

Comment:
Thinking aloud - start is the position from where we start downloading?

If yes then we need document about the Offset, that it keeps the downloaded data till Offset - 1.

Reply:
offset stores length up to which data is downloaded. So, an offset of 5 MiB means 5 MiB of data is downloaded.

start is the position(starts from 0) from which the next download should start. So, initially start = 0, limit = 8MiB, offset = 0. When 8 MiB is downloaded, both start and offset becomes 8MiB and limit becomes 16 MiB.

Does the above need to change ? Maybe I am missing something here?

Reply:
Got it. But I was thinking offset as index till data has been downloaded. In general also offset treated as index not as length. No, the only thing we need document about the offset properly, as it contains data [0, offset).